### PR TITLE
Cylinder raytracing

### LINKED ISSUE #115: Seperate classes for `receiver` and `calibration_target`
**Problem**
Right now both the calibration targets and the receiver are modeled as a flat plane. Therefore they are simulated by the same class: `TargetArea`. Real receivers can have unique geometries, that are sometimes not just a flat plane.

**Solution**
Once we use more accurate receivers we should create seperate classes for calibration targets, that are usually flat and receivers that probably are not flat.

### PR DESCRIPTION
**Pull request by MarleneBusch**
# Description

We implemented cylinder raytracing which can be used in combination with the earlier ray tracing.
To achieve this we implemented a new target area type and target area structure altogether. We had to make many changes in many files to adapt to this new structure.
We also fixed some bugs in the ray blocking. 

Fixes #115 

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: examples/field_optimizations/generate_viable_heliostats_list.py [hunk 1]
```diff
@@ -1,55 +1,28 @@
...
 import yaml
 
-from artist.util.environment_setup import get_device
-
 
 def find_viable_heliostats(
     data_directory: pathlib.Path,
     minimum_number_of_measurements: int,
-    kinematics_reconstruction_image_type: str,
+    kinematic_reconstruction_image_type: str,
```
Comment (mcw92) [inline examples/field_optimizations/generate_viable_heliostats_list.py]: Is there a reason why this file uses "kinematic" instead of "kinematics" everywhere?

File: tests/scene/test_light_source_array.py [hunk 2]
```diff
@@ -1,23 +1,21 @@
+from unittest import mock
```
Comment (mcw92) [inline tests/scene/test_light_source_array.py:1]: Just asking out of curiosity: Why did you switch from `pytest_mock` to `unittest.mock`?

File: artist/core/surface_reconstructor.py [hunk 3]
```diff
@@ -278,30 +308,42 @@ def reconstruct_surfaces(
...
+                    torch.tensor(
+                        [self.optimizer_dict[config_dictionary.max_epoch]],
+                        device=device,
+                    )
                 )
+
+                # Optimization loop.
                 while (
-                    loss > float(self.optimizer_dict[config_dictionary.tolerance])
+                    total_loss > float(self.optimizer_dict[config_dictionary.tolerance])
```
Comment (mcw92) [inline artist/core/surface_reconstructor.py:337]: I think this should use `total_loss` instead of `loss`. In this method, `loss` was initialized to `torch.inf` before and never updated, so using it here makes the stopping logic ineffective. This line should track the actual optimized objective (`total_loss`). I also initialized the `total_loss` as `+inf` (instead of `loss`) and removed `loss` completely.
@MarleneBusch can you please double-check whether that makes sense?

File: artist/field/solar_tower.py [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+            cylindrical = cast(
+                TowerTargetAreasCylindrical,
+                self.target_areas[index_mapping.cylindrical_target_areas],
+            )
+            centers[~planar_mask] = (
+                cylindrical.centers[cylinder_indices]
+                + cylindrical.radii[cylinder_indices][:, None]
+                * cylindrical.normals[cylinder_indices]
+            )
+        centers[:, 3] = 1.0
```
Comment (mcw92) [inline artist/field/solar_tower.py:189]: `centers[:, 3] = 1.0` was inside the cylindrical branch, so if all requested targets are planar, the homogeneous coordinate would remain 0. I moved this assignment outside the conditional so the 4th dimension is always set correctly for all returned centers.

File: artist/scenario/scenario.py [hunk 5]
```diff
@@ -199,23 +196,22 @@ def load_scenario_from_hdf5(
...
-                    "There is an error in the prototype. When using the rigid body kinematics, all actuators for this prototype must have the same type."
-                )
-            else:
-                prototype_actuator_type = prototype_actuator_type_list[0]
+        unique_prototype_actuators = set(prototype_actuator_type_list)
+
+        if len(unique_prototype_actuators) > 1:
+            raise ValueError("Prototype actuators must all have the same type.")
+
+        prototype_actuator_type = prototype_actuator_type_list[0]
```
Comment (mcw92) [inline artist/scenario/scenario.py:214]: I removed the rigid-body-only guard here and enforce a single prototype actuator type for all kinematics types because `prototype_actuator_type` was used unconditionally below afterwards and ARTIST currently only implements rigid body kinematics anyway. This avoids a potential unbound variable. @MarleneBusch Please double-check.

File: artist/util/utils.py [hunk 6]
```diff
@@ -528,58 +551,161 @@ def get_center_of_mass(
...
+        radii = cylindrical.radii[cylinder_indices].flatten()
+
+        theta = (
+            bitmap_coordinates[~planar_mask, index_mapping.unbatched_bitmap_e]
+            / (bitmap_resolution[index_mapping.unbatched_bitmap_e] - 1)
+            - 0.5
+        ) * cylindrical.opening_angles[cylinder_indices].flatten()
+        z = (
+            (
+                (bitmap_resolution[index_mapping.unbatched_bitmap_u] - 1)
```
Comment (mcw92) [inline artist/util/utils.py:689]: I changed the cylindrical `z` mapping to use `bitmap_resolution[unbatched_bitmap_u]` (instead of `unbatched_bitmap_e`) in the numerator. @MarleneBusch Could you please double-check this axis choice? I believe the previous version mixed E/U resolutions and would distort vertical mapping when `res_e != res_u`.

File: tests/field/test_tower_target_areas_cylindrical.py [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+    mock_normal.__getitem__.return_value = [0.0, 1.0, 0.0, 0.0]
+
+    mock_radius = mock.MagicMock()
+    mock_radius.__getitem__.return_value = 4.0
+
+    mock_height = mock.MagicMock()
+    mock_height.__getitem__.return_value = 6.0
+
+    mock_opening_angle = mock.MagicMock()
+    mock_opening_angle.__getitem__.return_value = 180.0
```
Comment (mcw92) [inline tests/field/test_tower_target_areas_cylindrical.py]: Is this in degree? I think the class actually expects the opening angle in radians (that's at least what the docstring says).

**Review by mcw92 — APPROVED** [review]
I carefully double-checked the code in the actual package, the corresponding tests, and updates in the docs pages. I made some in-line comments where feedback/explicit approval is required @MarleneBusch. Otherwise, I think we are ready to merge. Please correct me if I missed something important @MarleneBusch. 

Things that need to be considered afterwards:
- Check final `generate_results.py` and `generate_plots.py` scripts.
- Find a new place for `examples/`
- Check scripts in `tutorials/`.
- Maybe one more round of docstrings with Cline.
- Ask Cline how to improve the package structure, in particular w.r.t. `utils/`.

File: examples/field_optimizations/generate_viable_heliostats_list.py (hunk 1 above)
Comment (MarleneBusch) [reply examples/field_optimizations/generate_viable_heliostats_list.py]: no i fixed it now

File: tests/scene/test_light_source_array.py (hunk 2 above)
Comment (MarleneBusch) [reply tests/scene/test_light_source_array.py:1]: inbetween i thought i needed the more general unittest.mock in another test file, then i switched all tests to unittest.mock, to be consistent and then i realized that my assumption in the initial test was wrong and that i didnt need the switch at all but was too lazy to change it back. - not really a good reason.

File: artist/core/surface_reconstructor.py (hunk 3 above)
Comment (MarleneBusch) [reply artist/core/surface_reconstructor.py:337]: yes that makes sense, thanks

File: artist/field/solar_tower.py (hunk 4 above)
Comment (MarleneBusch) [reply artist/field/solar_tower.py:189]: This works.
Although i thought the previous version also works with one less operation, if all target areas are planar.

In the planar case:
`centers[planar_mask] = planar.centers[target_area_indices[planar_mask]]`
and the `planar.centers` attribute is already always guranteed to be homogeneous coordinates.

I might be mistaken but i dont care which version we take so lets take yours?

File: artist/scenario/scenario.py (hunk 5 above)
Comment (MarleneBusch) [reply artist/scenario/scenario.py:214]: To be honest I am not sure if I fully understand the change here, but since there is a test that checks this section and everything still runs, I would trust your new version, we can also talk about it tomorrow if you wish

File: artist/util/utils.py (hunk 6 above)
Comment (MarleneBusch) [reply artist/util/utils.py:689]: looks like the correct fix to me. thanks

File: tests/field/test_tower_target_areas_cylindrical.py (hunk 7 above)
Comment (MarleneBusch) [reply tests/field/test_tower_target_areas_cylindrical.py]: Yes your observation is correct. The test is not actually wrong, just misleading, this test just checks if values are loaded correctly, so it would think that 180 is radians which is a valid value, just very much unlikley but it doesnt matter here. I changed it to pi anyways :)

**Comment by MarleneBusch** [comment]
I agree to a merge now. Thanks for all your work Marie.
